File: orion-build-svc-go/internal/models/build_cache.go
package models

// ==================== Build Cache (canonical types in models.go) ====================
// BuildCacheConfig, CacheEntry, CreateBuildCacheConfigInput,
// UpdateBuildCacheConfigInput are defined in models.go.
//
// Additional domain-only types below — these are NOT in models.go because
// they are only used by the build cache service.

// CacheLevel defines the scope of a cache configuration.
type CacheLevel string

const (
	CacheLevelGlobal   CacheLevel = "global"
	CacheLevelPipeline CacheLevel = "pipeline"
	CacheLevelTask     CacheLevel = "task"
)

// CacheStorageType defines where cache data is stored.
type CacheStorageType string

const (
	CacheStorageTypeLocalVolume CacheStorageType = "local-volume"
	CacheStorageTypeS3          CacheStorageType = "s3"
	CacheStorageTypeNFS         CacheStorageType = "nfs"
)

// CacheCleanupPolicy defines the eviction strategy for cache entries.
type CacheCleanupPolicy string

const (
	CacheCleanupPolicyLRU    CacheCleanupPolicy = "lru"
	CacheCleanupPolicyTTL    CacheCleanupPolicy = "ttl"
	CacheCleanupPolicyManual CacheCleanupPolicy = "manual"
	CacheCleanupPolicyNever  CacheCleanupPolicy = "never"
)

// CacheStatus defines whether caching is enabled or disabled.
type CacheStatus string

const (
	CacheStatusEnabled  CacheStatus = "enabled"
	CacheStatusDisabled CacheStatus = "disabled"
)

// ListCacheConfigsOptions filters cache configs.
type ListCacheConfigsOptions struct {
	Level  CacheLevel
	Status CacheStatus
	Limit  int
	Offset int
}

// ListCacheEntriesOptions filters cache entries.
type ListCacheEntriesOptions struct {
	ConfigID string
	Limit    int
	Offset   int
}

File: blueprints/orion-code-svc/src/models/BuildCache.ts
/**
 * Build Cache Models - 构建缓存数据模型
 */

export enum CacheLevel {
  GLOBAL = 'global',
  PIPELINE = 'pipeline',
  TASK = 'task',
}

export enum CacheStatus {
  ENABLED = 'enabled',
  DISABLED = 'disabled',
}

export enum CacheCleanupPolicy {
  LRU = 'lru',
  TTL = 'ttl',
  MANUAL = 'manual',
}

export enum CacheStorageType {
  LOCAL = 'local',
  S3 = 's3',
  NFS = 'nfs',
}

export interface BuildCacheConfig {
  id: string;
  level: CacheLevel;
  targetId?: string;
  status: CacheStatus;
  storageType: CacheStorageType;
  storagePath?: string;
  maxTotalSize?: number;
  maxAgeDays?: number;
  cleanupPolicy: CacheCleanupPolicy;
  cacheKeyPattern: string;
  cachePaths: string[];
  description?: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface CacheEntry {
  id: string;
  configId: string;
  cacheKey: string;
  hash: string;
  size: number;
  storagePath: string;
  hitCount: number;
  lastHitAt?: Date;
  expiresAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

export interface BuildCacheConfigCreateInput {
  level: CacheLevel;
  targetId?: string;
  status?: CacheStatus;
  storageType?: CacheStorageType;
  storagePath?: string;
  maxTotalSize?: number;
  maxAgeDays?: number;
  cleanupPolicy?: CacheCleanupPolicy;
  cacheKeyPattern?: string;
  cachePaths: string[];
  description?: string;
}

export interface BuildCacheConfigUpdateInput {
  status?: CacheStatus;
  storageType?: CacheStorageType;
  storagePath?: string;
  maxTotalSize?: number;
  maxAgeDays?: number;
  cleanupPolicy?: CacheCleanupPolicy;
  cacheKeyPattern?: string;
  cachePaths?: string[];
  description?: string;
}

export function createBuildCacheConfig(input: BuildCacheConfigCreateInput): BuildCacheConfig {
  const now = new Date();
  return {
    id: `cache-config-${Date.now()}`,
    level: input.level,
    targetId: input.targetId,
    status: input.status ?? CacheStatus.ENABLED,
    storageType: input.storageType ?? CacheStorageType.LOCAL,
    storagePath: input.storagePath,
    maxTotalSize: input.maxTotalSize,
    maxAgeDays: input.maxAgeDays,
    cleanupPolicy: input.cleanupPolicy ?? CacheCleanupPolicy.LRU,
    cacheKeyPattern: input.cacheKeyPattern || '{{level}}-{{targetId}}-{{hash}}',
    cachePaths: input.cachePaths,
    description: input.description,
    createdAt: now,
    updatedAt: now,
  };
}

export function updateBuildCacheConfig(config: BuildCacheConfig, input: BuildCacheConfigUpdateInput): BuildCacheConfig {
  return { ...config, ...input, updatedAt: new Date() };
}

export function createCacheEntry(configId: string, hash: string, storagePath: string): CacheEntry {
  const now = new Date();
  return {
    id: `cache-entry-${Date.now()}`,
    configId,
    cacheKey: '',
    hash,
    size: 0,
    storagePath,
    hitCount: 0,
    createdAt: now,
    updatedAt: now,
  };
}

export function generateCacheKey(pattern: string, hash: string): string {
  return pattern.replace('{{hash}}', hash);
}
